File: frontend/src/gl/hex.ts
// Flat-top odd-q hex math for the WebGL layer, MIRRORING engine/space/navmap hex.go
// (navmap is the authority — the frontend never hardcodes the convention elsewhere).
// Kept separate from render/terrain.ts (that file is the pure-2D raster path); this
// one adds the pixel→cell inverse the 3D layer needs to seat entities on the terrain.

const SQRT3 = Math.sqrt(3)

// World centre of offset(col,row) for a flat-top odd-q grid whose offset(0,0) centre
// is world (0,0). EXACTLY navmap's offsetToAxial + hexToPixel (matches render/terrain.ts).
export function hexCentre(col: number, row: number, size: number): { x: number; y: number } {
  const q = col
  const r = row - (col + (col & 1)) / 2
  return { x: size * 1.5 * q, y: size * (SQRT3 / 2 * q + SQRT3 * r) }
}

// world (x,y) → nearest offset (col,row): inverse of hexCentre (axial + cube rounding).
export function pixelToOffset(x: number, y: number, size: number): { col: number; row: number } {
  const q = (2 / 3) * (x / size)
  const r = (-1 / 3) * (x / size) + (SQRT3 / 3) * (y / size)
  // cube round (x=q, z=r, y=-x-z)
  const cy = -q - r
  let rx = Math.round(q), ry = Math.round(cy), rz = Math.round(r)
  const dx = Math.abs(rx - q), dy = Math.abs(ry - cy), dz = Math.abs(rz - r)
  if (dx > dy && dx > dz) rx = -ry - rz
  else if (dy > dz) rz = -rx - ry // (ry unused after this)
  return { col: rx, row: rz + (rx + (rx & 1)) / 2 }
}

// The 6 hex edges' axial neighbour steps (dq,dr), ordered to match the prism template's
// edge midpoints (edge i midpoint direction = (i+0.5)·60°). Edge i's wall drops toward
// this neighbour, and (i+3)%6 is its reciprocal (so shared walls collapse — no z-fight).
export const HEX_DIRS: ReadonlyArray<readonly [number, number]> = [
  [1, 0], [0, 1], [-1, 1], [-1, 0], [0, -1], [1, -1],
]

export function neighbourOffset(col: number, row: number, k: number): { col: number; row: number } {
  const ar = row - (col + (col & 1)) / 2
  const nc = col + HEX_DIRS[k][0]
  const nr = ar + HEX_DIRS[k][1]
  return { col: nc, row: nr + (nc + (nc & 1)) / 2 }
}

File: backend/engine/space/navmap/hex.go
package navmap

import "math"

// Flat-top hexagonal grid geometry (docs/hex-grid.md, phase H2). Pure functions over axial (q,r) hex
// coordinates, an offset(col,row) rectangular layout, and continuous world points measured relative to
// a grid origin. navmap OWNS the hex convention: H3 wires CellOf/CellCenter/Neighbors onto these
// helpers; pathfind + the frontend consume it and never redefine it. Orientation is FLAT-TOP.
//
// Conventions (size = circumradius = hex centre→vertex distance, world units):
//   - Axial neighbour topology is orientation-independent; flat-top only changes the pixel mapping.
//   - The offset layout is "odd-q": odd columns are shifted DOWN half a row (flat-top). This is the
//     bijection between the fixture/wire/frontend rectangular array and internal axial coords.
//   - No RNG, no wall-clock — deterministic pure math (D12). Negative-safe (floor-free; cube rounding).

// sqrt3 is √3, computed once (Go's math package has Sqrt2 but not Sqrt3).
var sqrt3 = math.Sqrt(3)

// hexDirs are the 6 axial neighbour directions in FIXED canonical order — the single neighbour
// authority (navmap.Neighbors returns these; pathfind must not hardcode its own set).
var hexDirs = [6][2]int{
	{+1, 0}, {+1, -1}, {0, -1}, {-1, 0}, {-1, +1}, {0, +1},
}

// hexToPixel maps axial (q,r) to the continuous centre of that hex, relative to the grid origin
// (the caller adds the world origin). Flat-top.
func hexToPixel(q, r int, size float64) (x, y float64) {
	fq, fr := float64(q), float64(r)
	x = size * (1.5 * fq)
	y = size * (sqrt3/2*fq + sqrt3*fr)
	return
}

// pixelToHex maps a continuous point (relative to the grid origin) to the axial (q,r) of the hex
// containing it: flat-top pixel→hex + cube rounding. Deterministic; negative-safe.
func pixelToHex(x, y, size float64) (q, r int) {
	fq := (2.0 / 3.0 * x) / size
	fr := (-1.0/3.0*x + sqrt3/3.0*y) / size
	return cubeRound(fq, fr)
}

// cubeRound rounds a fractional axial (qf,rf) to the nearest hex via the cube constraint x+y+z=0,
// re-deriving the component with the LARGEST rounding error (the standard robust hex rounding — doing
// this in axial directly is the classic bug source, hence the cube detour + this dedicated test).
func cubeRound(qf, rf float64) (q, r int) {
	xf, zf := qf, rf
	yf := -qf - rf
	rx := math.Round(xf)
	ry := math.Round(yf)
	rz := math.Round(zf)
	xd, yd, zd := math.Abs(rx-xf), math.Abs(ry-yf), math.Abs(rz-zf)
	switch {
	case xd > yd && xd > zd:
		rx = -ry - rz
	case yd > zd:
		ry = -rx - rz
	default:
		rz = -rx - ry
	}
	return int(rx), int(rz) // q = cube.x, r = cube.z
}

// hexDistance is the number of single-step moves between two axial hexes (the hex metric).
func hexDistance(q1, r1, q2, r2 int) int {
	return (hexAbs(q1-q2) + hexAbs(q1+r1-q2-r2) + hexAbs(r1-r2)) / 2
}

// axialToOffset / offsetToAxial convert between axial (q,r) and the offset(col,row) rectangular layout
// (flat-top odd-q). Bijection; numerator (q + (q&1)) is always even so the /2 is exact for any sign.
func axialToOffset(q, r int) (col, row int) {
	return q, r + (q+(q&1))/2
}

func offsetToAxial(col, row int) (q, r int) {
	return col, row - (col+(col&1))/2
}

func hexAbs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
